File: clj-advoc/src/clj_advoc/day_6.clj
(ns clj-advoc.day-6
  (:require [clojure.java.io :as io]
            [clojure.set :as cset]
            [clojure.string :as s]))


(defn problem-input
  [filename]
  (let [rdr (io/reader (io/resource filename))]
    (s/split (slurp rdr) #"\n\n")))

;; ────────────────────────────── PART 1 ──────────────────────────────

(defn count-qs-part-1
  "Given a list of replies by all the people in the group,
  return the count of yes replies by ANYONE in the group."
  [group]
  (-> group
      set
      count))


(defn solution-part-1
  [groups]
  (->> groups
       (map #(s/replace % "\n" ""))
       (map count-qs-part-1)
       (reduce +)))

;; ────────────────────────────── PART 2 ──────────────────────────────

(defn count-qs-part-2
  "Given a list of replies by all the people in the group,
  return the count of yes replies by everyone in the group."
  [group]
  (let [per-reply (s/split group #"\n")]
    (count (reduce (fn [it rep]
                     (cset/intersection it (set rep)))
                   (set (first per-reply))
                   (rest per-reply)))))

(defn solution-part-2
  [groups]
  (->> groups
       (map count-qs-part-2)
       (reduce +)))

(defn main
  []
  (let [input (problem-input "day-6-part-1.input")]
    (-> input
        (solution-part-1)
        (println "<-- Answer"))
    (-> input
        (solution-part-2)
        (println "<-- Answer"))))

File: py-advoc/day_6.py
import os

INPUT_PATH = "../inputs/"


def problem_input(filename):
    path = os.path.join(INPUT_PATH, filename)
    inputs = []
    with open(path, 'r') as f:
        return f.read().split('\n\n')


def count_qs_part_1(group):
    """
    Given a list of replies by all the people in the group,
    return the count of yes replies by ANYONE in the group.
    """
    return len(set(group.replace('\n', '')))

def solution_part_1(groups):
    return sum(map(count_qs_part_1, groups))


def count_qs_part_2(group):
    """
    Given a list of replies by all the pople in the group,
    return the count of yes replies by EVERYONE in the group.
    """
    # Always beware of a trailing newline in python ( •̀ᴗ•́ )و ̑̑
    replies = group.strip('\n').split('\n')
    int_set = set(replies[0])
    for rep in replies[1:]:
        int_set = int_set.intersection(rep)

    return len(int_set)

def solution_part_2(groups):
    return sum(map(count_qs_part_2, groups))


if __name__ == "__main__":
    inpt = problem_input("day-6-part-1.input")
    ans = solution_part_1(inpt)
    print(f'Answer for Part 1: {ans}')

    ans = solution_part_2(inpt)
    print(f'Answer for Part 2: {ans}')
